man/pulsar_deliver.8.man
.TH pulsar_deliver 8 "10 July 2003"

.SH "NAME"
pulsar_deliver - pulsar delivery agent

.SH "SYNOPSIS"
pulsar [-h] [-d[num]] [-rrealm] username

.SH "DESCRIPTION"
A delivery agent from your SMTP server package 
.I can
be used to perform delivery.

Pulsar_deliver however knows how to process /etc/pulsar.conf, knows
wich gid/uid to use on delivery and it uses the same kind of locking
as pulsar POP3 daemon. Pulsar_deliver should be run as user 
.B root
in most conditions so it can perform thee needed uid/gid change before
delivery. 
.I Security_note:
If you do delivery to a mailspool and (virtual or system) users only
access their mail thru pulsar POP3 daemon, you can make all of the
mailspool owned by a system account. This way pulsar_deliver
can be run under this system account and no uid/gid change is necessary.

Pulsar mbox and Maildir files are compatible with standard mbox and Maildir
formats. However to speed-up opening of a mailbox certain information
is cached. Pulsar_deliver knows how to update cache information. If other
delivery agents are used cache information needs to be updated when user
tries to access his mail and this can take some time.

For mbox pulsar uses POSIX fnctl() locking mechanism that
.I should
also work over NFS. If you plan to do delivery over an NFS netowork
.B make sure
that your NFS server 
.B and
client support POSIX locks.
.I Note:
Maildir format doesn't need locking, so this it not an issue.

.B Since there weren't many requests delivery to mbox format is currently not implemented.

.dnl Pulsar_deliver uses the correct multilevel >From_ quoting.

.TP 
.B -h
.sp
prints help message.

.TP
.B -d
num
.sp
.I num
specifies the desired debug level.

.TP
.B -r
realm
.sp
Your SMTP daemon will know for wich realm delivery should be made.

.TP
.B username
.sp
specifies for which user delivery should be done.

.SH "POSTFIX"
First try to use postfixs own delivery agent. It is flexible enough to be used
for most of the configurations. Nothing special needs to be configured for this.

.B /etc/pulsar.conf
.nf
	mailspool = /var/spool/mail:0600

.SH "POSTFIX SYSTEM ACCOUNT DELIVERY"
In this example mailspool is ownder by a system account with uid =
.B mail
and gid =
.B mail.
Users access their mail only with POP3. We use Maildir system spool
delivery.

.B /etc/postfix/master.cf
.nf
	pulsar	unix	-	n	n	-	1	pipe flags=FR user=mail:mail argv=/usr/sbin/pulsar_deliver ${user}

.fi
Delivery is done to default realm. Add the -r option to do delivery to a
non-default realm. The F and R flags indicate that postfix should
prepend envelope header and Return-Path field to a message.

.B /etc/postfix/main.cf
.nf
	pulsar_destination_recipient_limit = 1
	mailbox_transport = pulsar

.B /etc/pulsar.conf
.nf
	mailspool = /var/spool/Maildir/:0600
	mailspool_owner = mail:mail

.SH "FULLY FLEXIBLE POSTFIX DELIVERY"
Postfix at the time of this writing has a policy not to run any of non-postfix
utilities as superuser.
The supplied patch in ./patches/postfix-*/transport_allow_root_delivery.diff
removes this restriction in the least intrusive way. It will add support for
"transport_allow_root_delivery" option in postfix main.cf configuration file.

.I transport
is the name of the delivery agent as specified in postfix master.cf file that
should be allowed to run as root.

.B /etc/postfix/master.cf
.nf
	pulsar	unix	-	n	n	-	1	pipe flags=FR user=root argv=/usr/sbin/pulsar_deliver -r test ${user}

.fi
With this line postfix will always deliver to realm "test". To make delivery to
default realm remove the -r option. The F and R flags indicate that postfix should
prepend envelope header and Return-Path field to a message.

.B /etc/postfix/main.cf
.nf
	pulsar_destination_recipient_limit = 1
	pulsar_allow_root_delivery = yes
.fi

Where
.I pulsar
is the transport agent as specified in the first column of master.cf file.
Destination recepient limit must be set to 1 as pulsar_deliver must be called once
for every user.

.SH "SEE ALSO"
pulsar(8), pulsar.conf(5), pulsar_crpyt(1), mbox(5), maildir(5), pipe(8).


doc/CHANGES
2003/07/11 - version 0.1.1:
 - Fixed numerous Maildir bugs
 - Added APOP support
 - pulsar_crypt can also calculate plain (not salted) md5 hashes
   (Usefull for APOP debugging)

2003/07/10 - version 0.1.0:
 - error_facility now loggs application returned error string/rc
   and loggs exact point of error origin
 - added maildir support
 - reactivated a build of pulsar_deliver
 - enhancements to build system
 - epm multiple package list file (rpm, deb, ...)
 - added md5hash as UIDL in Maildir mailstore
 - updated pulsar.conf and pulsar_deliver man pages

2002/10/09 - version 0.0.9:
 - updated pulsar.conf(5) man page
 - FreeBSD logging fixes (logging using mail.info)
 - IMP bugfix
 - FreeBSD makefile fixes
 - autoconf improvements
 - MySQL password column defaults to 'plaintext' passwords
   if hash type is not specified.

2002/08/11 - version 0.0.8:
 - fixed compile error in MySQL user authentication code
 - ported to FreeBSD 4.6-STABLE

2002/08/06 - version 0.0.7:
 - implemented assert with syslog output and ifdebug() conditional compile
 - mailspool_owner now overrides any user-db provided information
 - moved all the sources to ./src subdirectory and updated the
   build mechanism
 - rewrote the mailbox handling code (delivery isn't implemented yet)
 - added CAPA POP3 command
 - added SSL support
 - added ssl_listen_on option
 - added certificate option
 - fixed a bug in IP based virtual where interface sometimes
   isn't matched.
 - fixed an auth_db = file:.. bug when plaintext was not enabled
   if hash_type_list wasn't specified.
 - FreeBSD portability changes

2002/04/21 - version 0.0.6:
 - bugfix: cache-handling bug

2002/04/11 - version 0.0.5:
 - numerous mailbox code updates and fixes
 - added caching support for faster loading of huge mailbox files
 - added UIDL support
 - MySQL authentication now supports plaintext and crypt hashed
   passwords (use pulsar_crypt for creating password hashes).
 - Manualy specify password hash type in auth_db=file option and
   in password_column mysql option. Defaults to plaintext if
   omitted. Supported hash types are plaintext and crypt (system
   provided DES and MD5).

2002/03/24 - version 0.0.4:
 - updated mailbox code (buffering and optimization)
 - cleanups in MySQL authentication code
 - fixed pulsar_deliver utility

2002/03/14 - version 0.0.3:
 - added native mysql support
 - added mailspool_owner option and removed
   mailspool_gid
 - expanded the mailspool option to include permissions
 - code cleanups in config file code
 - listening sockets now have SO_REUSEADDR set
 - implemented TOP command
 - rewrote mailbox handling code becouse of Outlook
   problems

 KNOWN BUGS:
 - pulsar_deliver is broken, don't use it
 - new mailbox handling code is not optimized yet

2002/02/16 - version 0.0.2:
 - fixed segmentation fault at client authentication when
   realm_chars option isn't used
 - PAM authentication musn't try to retreive users homedir
   information when doing mailspool delivery
 - Mailspool files should be created with permissions 660

2002/01/29 - version 0.0.1:
 - initial alpha release


src/pulsar_deliver.c
/*
 * Copyright (C) 2001 Rok Papez <rok.papez@lugos.si>
 * Rok Papez
 * Hribovska pot 17
 * 1231 Ljubljana - Crnuce
 * EUROPE, Slovenia
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#ifdef   HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <pwd.h>
#include <grp.h>
#include <sys/types.h>

#include "pulsar.h"
#include "users.h"
#include "error_facility.h"

strStaticData g;
//------------------------------------------------------------------------------------------------------------
void help(const char *filename) {
  fprintf(stderr,
          "Usage: %s [-h] [-d[num]] [-rrealm] username\n"
          "\t-rrealm\tDeliver to user 'username' in realm 'realm'\n"
          "\t-dnum\tEnable debugging. Optional parameter num \n"
          "\t\tspecifies debug level (default = 1).\n"
          "\t-h\tdisplay this help message.\n"
          , filename
         );
  return;
}
//------------------------------------------------------------------------------------------------------------
int main(int argc, char *argv[]) {
  int rc = EX_OK;

  if(argc<2) {
    fprintf(stderr,"ERROR: This command should not be run by hand!\n");
    return EX_USAGE;
  }

  err_init_app("pulsar_deliver");
  err_set_debug_level(defDefaultDebugLevel);

  // read in the options from config file.
  g.cfg = cfg_read_config(defConfigFile);
  if (!g.cfg) {
    fprintf(stderr,"%s\n", err_get_error());
    err_close();
    return err_get_rc();
  }
  err_set_debug_level(g.cfg->debug);
  g.realm = get_realm(NULL, g.cfg); // select default realm!
  if (!g.realm) {
    err_internal_error();
    goto error;
  }

  // read options from command line
  opterr = 0;
  while(1) {
    rc = getopt(argc,argv,"hd::r:");
    if(-1 == rc) // no more options
      break;
    switch(rc){
    case 'r':
        g.realm = get_realm(optarg, g.cfg); // select default realm!
        if (!g.realm) {
          err_error(EX_DATAERR,"Realm \"%s\" not found\n",optarg);
          goto error;
        }
        err_debug(5, "Realm \"%s\" located", optarg);
        break;
    case 'd':
        if(optarg)
          err_set_debug_level(atoi(optarg));
        else
          err_set_debug_level(defDefaultDebugLevel);
        break;
    case '?':
        fprintf(stderr,"ERROR: unknown switch -%c\n",optopt);
    case 'h':
    default:
        help(argv[0]);
        err_error(EX_USAGE, "");
        goto error;
    }
  }
  if(argc-1 != optind) {
    fprintf(stderr,"ERROR: This command should not be run by hand!\n");
    err_error(EX_USAGE, "Insufficient parameters");
    goto error;
  }

  g.user = argv[optind];
  err_debug(3, "Delivery for user \"%s\"", g.user);

  // Get's user data, drops privs and open mailstore
  rc = users_getinfo();
  err_debug(0, "getinfo rc = %d", rc);
  if(defOK != rc) // all getinfos have to set err_error.
    goto error;   // TODO: shouldn't we use def* errors ? We have to unify error reports.

  // do delivery
  rc = mailstore_deliver(g.head, 0); // read from stdin
  if(-1 == rc) {
    mailstore_close(g.head, 0); // try to cleanup what we can...
    goto error;
  }

  // close mailbox, free options, do a nice cleanup.
  rc = mailstore_close(g.head, 0);
  if(-1 == rc)
    goto error;
  free_all_options(&g.cfg);
  err_close();
  return EX_OK;

error:
  fprintf(stderr, "%s\n", err_get_error());
  free_all_options(&g.cfg);
  err_close();
  return err_get_rc();
}
//------------------------------------------------------------------------------------------------------------
